### ISSUE TITLE
SEssion State locked?

### ISSUE THREAD

**Issue by jetic83** [DavidLievrouw/Nancy.Session.InProc#1]
Thank you for this package!
Is the memory of the session locked during a request? So, when multiple queries of one user arrive, do these queries have to wait one after the other?

**Comment by DavidLievrouw** [comment]
Hi,

Not entirely sure what you mean with "the memory". But in any case:
- The only thread-safe aspect of this library is in the acquisition and saving of the Session instance, as shown here: [InProcSessionCache.cs](https://github.com/DavidLievrouw/Nancy.Session.InProc/blob/master/src/Nancy.Session.InProc/InProcSessionsManagement/Cache/InProcSessionCache.cs)
- As you know, requests are handled in separate threads, so no, in that way, requests do not have to wait for one another. All concurrent requests have access to the session data, at the same moment in time.
- The data in the ISession itself is an instance of the default Session implementation of NancyFx [Session.cs](https://github.com/NancyFx/Nancy/blob/v1.4.4/src/Nancy/Session/Session.cs) and uses a Dictionary internally. So that is not thread-safe. [InProcSessionManager.cs, line 46](https://github.com/DavidLievrouw/Nancy.Session.InProc/blob/master/src/Nancy.Session.InProc/InProcSessionsManagement/InProcSessionManager.cs)

So in short:
- NOT thread-safe: The data in the session.
- Thread-safe: creating, loading and abandoning the reference to the actual Session.

Examples:
- Two initial requests from the same user arrive at the same time: You cannot have threading issues by two requests coming in concurrently, each creating its own session. Session creation, loading and disposal are thread-safe. In this example, there will be only one session for that user.
- Two requests change session data at the same time: The NancyFx implementation of ISession uses a regular Dictionary, so that is not thread-safe. Both requests will "see" each others changes to the session data.

Hope this helps.

**Comment by jetic83** [comment]
Thanks for this answer! I think you are right and since this Session state works with a dictionary in the global cache, there is no locking involved. 

I basically meant the problem which apparantly arises with typical ASP.net webapps, as described here
https://stackoverflow.com/questions/3629709/i-just-discovered-why-all-asp-net-websites-are-slow-and-i-am-trying-to-work-out
